extensions/websockets-next/deployment/src/main/java/io/quarkus/websockets/next/deployment/WebSocketProcessor.java

```diff
@@ -419,11 +425,33 @@ public void registerRoutes(WebSocketServerRecorder recorder, HttpRootPathBuildIt
             builder
                     .displayOnNotFoundPage("WebSocket Endpoint")
                     .handlerType(HandlerType.NORMAL)
-                    .handler(recorder.createEndpointHandler(endpoint.generatedClassName, endpoint.endpointId));
+                    .handler(recorder.createEndpointHandler(endpoint.generatedClassName, endpoint.endpointId,
+                            endpointToUpgradePolicy.get(endpoint.endpointId)));
             routes.produce(builder.build());
         }
     }
 
+    @BuildStep
+    UnremovableBeanBuildItem makeHttpUpgradePoliciesUnremovable() {
+        // we access the policies programmatically
+        return UnremovableBeanBuildItem.beanTypes(HTTP_UPGRADE_POLICY_NAME);
+    }
+
+    @BuildStep
+    List<ValidationPhaseBuildItem.ValidationErrorBuildItem> validateHttpUpgradePolicyNotRequestScoped(
+            ValidationPhaseBuildItem validationPhase) {
+        return validationPhase
+                .getContext()
+                .get(BEANS)
+                .stream()
+                .filter(b -> b.hasType(HTTP_UPGRADE_POLICY_NAME))
+                .filter(b -> BuiltinScope.REQUEST.getName().equals(b.getScope().getDotName()))
```

Comment:
Shouldn't we test if the scope is supported instead? I.e. `@Dependent`, `@Singleton` or `@ApplicationScoped`?

Reply:
I thought it's rather discrete set of 4 scopes (probably something custom is possible?), but sure, your proposal is a correct way to do this. I'll fix it, thanks.